ctf_framework/views/writeup.py

```diff
@@ -0,0 +1,101 @@
+from django.contrib.auth.decorators import login_required
+from .base_view import *
+from django.urls import reverse
+from django.http import HttpResponseForbidden
+from django.utils.safestring import mark_safe
+from markdown import markdown
+from bleach import clean
+
+markdown_tags = [
+    "h1", "h2", "h3", "h4", "h5", "h6",
+    "b", "i", "strong", "em", "tt",
+    "p", "br",
+    "span", "div", "blockquote", "code", "hr", "pre",
+    "ul", "ol", "li", "dd", "dt",
+    "img",
+    "a",
+]
+
+markdown_attrs = {
+    "img": ["src", "alt", "title"],
+    "a": ["href", "alt", "title"],
+    "*": ["class"],
+}
+
+@login_required()
+def index(request, challenge_id):
+    """List all writeups for a challenge."""
+
+    user = UserProfile.objects.get(user=request.user)
+    challenge = Challenge.objects.get(id=challenge_id)
+
+    if challenge in user.completed_challenges.all():
+        writeups = Writeup.objects.filter(challenge_id=challenge_id)
+        writeup = Writeup.objects.filter(user=user.user, challenge=challenge).first()
+        context = {
+            "challenge": challenge,
+            "writeups": writeups,
+            "writeup": writeup
+        }
+        return render(request, "writeup/index.html", context)
+    else:
+        return HttpResponseForbidden("You must solve the challenge first")
+
+
+
+@login_required()
+def show(request, writeup_id):
+    """View the page of a specific writeup."""
+
+    user = UserProfile.objects.get(user=request.user)
+    writeup = Writeup.objects.get(id=writeup_id)
+    challenge = Challenge.objects.get(id=writeup.challenge.id)
+    html = clean(
+        markdown(writeup.markdown, extensions=['markdown.extensions.fenced_code', 'markdown.extensions.codehilite']),
+        markdown_tags,
+        markdown_attrs
+    )
+
+    if challenge in user.completed_challenges.all():
+        context = {
+            "writeup": writeup,
+            "html": mark_safe(html)
+        }
+        return render(request, "writeup/show.html", context)
+    else:
+        return HttpResponseForbidden("You must solve the challenge first")
+
+
+@login_required()
+def edit(request, writeup_id):
+    """Edit the existing writeup for a given challenge."""
+
+    if request.method == "POST":
+        user = UserProfile.objects.get(user=request.user)
+        writeup = Writeup.objects.get(id=writeup_id)
+        challenge = Challenge.objects.get(id=writeup.challenge.id)
+
+        if user.user == writeup.user and challenge in user.completed_challenges.all():
+            writeup = Writeup.objects.get(id=writeup_id)
+            writeup.markdown = request.POST["markdown"]
```

Comment:
I'm fairly certain Django will protect against most malicious input, but is there any filtering that should be done before dumping the POST directly into the DB?

Reply:
The filtering needs to happen after the markdown -> html has occurred, which is currently happening using bleach when we render the write up.

When this is shown raw for editing it will all be HTML escaped so should be fine too.

Or do you mean making sure the post data is a string? Haven’t worked with django before so not sure how that works :)